### PR TITLE
Fix #53

### LINKED ISSUE #53: Add an error when the names of `vc` does not match that of the input data frame
``` r
library(R2spa)
data("PoliticalDemocracy", package = "lavaan")
mod2 <- "
  # latent variables
    ind60 =~ x1 + x2 + x3
    dem60 =~ y1 + y2 + y3 + y4
    dem65 =~ y5 + y6 + y7 + y8
"
fs_dat2 <- get_fs(PoliticalDemocracy, model = mod2, std.lv = TRUE)
ecov_fs <- attr(fs_dat2, "av_efs")
dimnames(ecov_fs) <- lapply(dimnames(ecov_fs),
                            FUN = \(x) paste0("bs_", x))
tspa(model = "dem60 ~ ind60
              dem65 ~ ind60 + dem60",
     data = fs_dat2,
     vc = ecov_fs)
#> Warning in split.default(loadings, factor(rep(var, each = len), levels = var)):
#> data length is not a multiple of split variable
#> Error in str2lang(x): <text>:1:3: unexpected '*'
#> 1: ~ *
#>       ^
```

<sup>Created on 2023-05-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by winniewytse**

### CODE REVIEW

File: R/tspa.R [hunk 1]
```diff
@@ -113,27 +114,50 @@
 
 
 tspa <- function(model, data, reliability = NULL, se = NULL,
-                 vc = NULL, cross_loadings = NULL, ...) {
+                 fsT = NULL, fsL = NULL, ...) {
+
+  if (nchar(model) == 0) {
```
Comment (marklhc) [inline R/tspa.R]: I'm thinking whether we need `model` to also be character, as `nchar(12)` for example would be 2

File: R/tspa.R [hunk 2]
```diff
@@ -113,27 +114,50 @@
...
+
   if (!is.null(reliability)) {
     stop("tspa() currently does not support reliability model")
   }
+
   if (!is.data.frame(se)) {
     se <- as.data.frame(as.list(se))
   }
-  multigroup <- nrow(se) == 1 | is.list(vc)
+  multigroup <- nrow(se) > 1 | is.list(fsT)
```
Comment (marklhc) [inline R/tspa.R]: @winniewytse could you also merge in the changes in the devel branch as now we use `se_fs` for the argument? Thank you.

File: R/tspa.R [hunk 3]
```diff
@@ -113,27 +114,50 @@
...
+
+  if (sum(is.null(fsT), is.null(fsL)) == 1) {
```
Comment (marklhc) [inline R/tspa.R]: You may check out the `xor` operator

File: R/tspa.R [hunk 4]
```diff
@@ -113,27 +114,50 @@
...
+  if (!is.null(fsT)) {
+    fs_names <- ifelse(multigroup, colnames(fsT[[1]]), colnames(fsT))
+    dat_names <- ifelse(multigroup, names(data[[1]]), names(data))
+    names_match <- lapply(fs_names, function(x) x %in% dat_names) |> unlist()
+    if (any(!names_match)) {
+      stop("Names of factor score variables do not match those in the input data.")
+    }
+  }
 
   if (multigroup) {
```
Comment (marklhc) [inline R/tspa.R:151]: Just checking: do we now use the multigroup function for everything?

File: vignettes/corrected-se.Rmd [hunk 5]
```diff
@@ -84,14 +84,14 @@ ev <- block_diag(attr(fs1, which = "fsT"),
                  attr(fs2, which = "fsT"))
 tspa_fit <- tspa(model = "dem60 ~ ind60",
                  data = fs_dat,
-                 cross_loadings = ld,
-                 vc = ev)
+                 fsL = ld,
+                 fsT = ev)
 # Unadjusted covariance
-vcov(tspa_fit)
+fsTov(tspa_fit)
```
Comment (marklhc) [inline vignettes/corrected-se.Rmd]: @winniewytse some of the names that contain `vc` (e.g., `vcov`) have been incorrectly changed. Could you fix those and run a check? Thanks.

File: vignettes/efa-score.Rmd [hunk 6]
```diff
@@ -116,8 +116,8 @@ Recover factor covariances with 2S-PA
 ts_fit <- tspa("",
```
Comment (marklhc) [inline vignettes/efa-score.Rmd:116]: Turns out I originally used the empty model (`""`) trick to get a default model with all covariances. Maybe we consider keeping this and instead check whether the model is a character variable? Thanks

File: tests/testthat/test-tspa.R [hunk 7]
```diff
@@ -456,3 +456,102 @@ test_that("Multiple-group multiple-factor example", code = {
...
+  expect_no_error(
+    tspa(model = "dem60 ~ ind60
+              dem65 ~ ind60 + dem60",
+         data = fs_dat2,
+         fsT = attr(fs_dat2, "fsT"),
+         fsL = attr(fs_dat2, "fsL"))
+  )
+})
+
+test_that("Test indicator names not starting with 'fs_'", {
```
Comment (marklhc) [inline tests/testthat/test-tspa.R:521]: Nice test. I would suggest also adding a quick one that using `fs_` or `bs_` give identical results.

**Review by marklhc — CHANGES_REQUESTED** [review]
Thanks Winnie. See my comments and let me know if you have any questions

File: R/tspa.R (hunk 4 above)
Comment (winniewytse) [reply R/tspa.R:151]: That's a good question. Yes but we haven't merged the changes in the `growth` branch to `devel`. I forsee there will be quite some conflicts to resolve and plan to complete this pull request before moving on to the pull request from `growth`.

File: vignettes/efa-score.Rmd (hunk 6 above)
Comment (winniewytse) [reply vignettes/efa-score.Rmd:116]: Sounds good. `tspa()` now only check if the model is a character variable.

**Comment by winniewytse** [comment]
Thank you Mark! Please review the changes and let me know if there are any further changes needed.

File: R/tspa.R (hunk 4 above)
Comment (marklhc) [reply R/tspa.R:151]: I see. Let's wait for that then.

File: R/tspa.R [hunk 8]
```diff
@@ -115,31 +116,54 @@
...
             "use `se_fs` instead.")
   }
+
   if (!is.data.frame(se_fs)) {
     se_fs <- as.data.frame(as.list(se_fs))
   }
-  multigroup <- nrow(se_fs) == 1 | is.list(vc)
+  multigroup <- nrow(se_fs) > 1 | is.list(fsT)
+
+  if (xor(is.null(fsT), is.null(fsL))) {
```
Comment (marklhc) [inline R/tspa.R:138]: I think it should actually be `!xor()`. I'll make the change.

**Pull request merged**